requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+- **Composable** - Query ClickHouse. Calculate bill. Charge stripe.
+- **Parallelize-able** - We should be able to run multiple bills at the same time
+- **Timeouts** - We should be able to set timeouts on steps
+- **Cancelable** - We should be able to cancel any billing gracefully
+- **Terminate-able** - We should be able to rip the carpet out from a billing routine
+- **Nonblocking failures** - A failure of one billing workflow should not halt the execution of others.
+
+### Celery tasks
+
+Our celery tasks are pretty diverse. We use it for collecting metrics, refreshing caches, and if you didn’t know - we used to even use it to process every single event 😱.
+
+What do we use Celery for now?
+
+In general they are all scheduled tasks:
+
+- Running queries and emitting metrics (statsd, prometheus)
+- Warming up caches
+
+Requirements:
+
+- **Retry-able** - If the thing fails, get back on that horse
+- **Resumable** - Some of these are heavy and we don’t want to requery CH
+- **Parallelize-able** - We should be able to run multiple tasks at once
+- **Timeouts** - We should be able to set timeouts on the task
+- **Terminate-able** - We should be able to rip the carpet out from under the Task
+- **Nonblocking failures** - A failure of one task should not halt the execution of others.
+
+### Historical Exports *(AKA Vanilla ETL)*
+
+We do a lot of moving around data. Actually, a good chunk of our business is just getting data form A to B safe and sound, and in on piece. Some of these jobs are short, some are long running. There are a few requirements and nice to haves around ETLs
+
+Requirements:
+
+- **Idempotent** or **exactly once** - no duplicates
+- **Retry-able** - If the thing fails, get back on that horse
+- **Resumable** - Nice to pick up where you left off
+- **DAG-able** - Most ETLs resemble a Directed Acyclical Graph. Move A → B → C
+    - You can think of this as **Composable**
+- **Parallelize-able** - We should be able to run multiple bits of the DAG at once
+- **Timeouts** - We should be able to set timeouts on the DAG and Nodes
+- **Cancelable** - We should be able to cancel the DAG’s execution *gracefully*
+- **Terminate-able** - We should be able to rip the carpet out from under the DAG
+- **Nonblocking failures** - A failure of one DAG should not halt the execution of others.
+
+### Webhooks
+
+Ahh webhooks. Webhooks are a particularly sticky problem.
+
+Requirements:
+
+- **Idempotent** or **exactly once** - no duplicates
+- **Retry-able** - If the thing fails, get back on that horse
+- **Resumable** - Helps with idempotency/exactly once
+- **Parallelize-able** - We should be able to run multiple webhooks at once and out of order (non-blocking)
+- **Timeouts** - We should be able to set timeouts for webhooks
+- **Terminate-able** - We should be able to rip the carpet out from under webhooks
+- **Nonblocking failures** - A failure of one webhook endpoint should not halt the execution of others.
+
```

Comment:
One difficulty with webhooks is the risk of being rate-limited by the destination, either on a `team*destination` basis, or even on a `destination, for all teams` basis. Most sources will nicely return a `429`, some might silently black-hole traffic.
Ideally, we would be able to rate-limit task execution rate on our end (by these dimensions), to buffer spikes and ease-in the volume when recovering from incident. We'd ideally have the choice of delaying (backfilling) or dropping (input spike).